scripts/fuzz_shell.js

```diff
@@ -243,19 +253,40 @@ var imports = {
     },
 
     // Export operations.
-    'call-export': (index) => {
-      callFunc(exportList[index].value);
+    'call-export': /* async */ (index) => {
+      /* await */ callFunc(exportList[index].value);
     },
-    'call-export-catch': (index) => {
-      return tryCall(() => callFunc(exportList[index].value));
+    'call-export-catch': /* async */ (index) => {
+      return tryCall(/* async */ () => /* await */ callFunc(exportList[index].value));
     },
 
     // Funcref operations.
-    'call-ref': (ref) => {
-      callFunc(ref);
+    'call-ref': /* async */ (ref) => {
+      if (JSPI) {
+        // This is a direct function reference, and just like an export, it must
+        // be wrapped for JSPI.
+        ref = wrapExportForJSPI(ref);
+      }
+      /* await */ callFunc(ref);
     },
-    'call-ref-catch': (ref) => {
-      return tryCall(() => callFunc(ref));
+    'call-ref-catch': /* async */ (ref) => {
+      if (JSPI) {
+        ref = wrapExportForJSPI(ref);
+      }
+      return tryCall(/* async */ () => /* await */ callFunc(ref));
+    },
+
+    // Sleep a given amount of ms (when JSPI) and return a given id after that.
+    'sleep': (ms, id) => {
+      if (!JSPI) {
+        return id;
+      }
+      return new Promise((resolve, reject) => {
+        setTimeout(() => {
+          resolve(id);
+        }, 0); // TODO: Use the ms in some reasonable, deterministic manner, but
+               //       the d8 shell ignores this anyhow.
```

Comment:
The sleep would lead to interleaved execution, though, which seems useful to test. That is, actual swapping back and forth of stacks.

Reply:
Returning a resolved promise would still interleave execution, though. It's basically equivalent to what d8 is doing since it's ignoring the sleep time anyway. For extra interleaving, rather than sleeping we could have an integer counter that recursively decrements on every turn of the microtask queue and finally resolves the promise when the counter hits zero.

Reply:
Hmm, yeah, we could manage interleaving manually that way, interesting idea. I can add that to the TODO